runtime/lua/vim/fs.lua

```diff
@@ -334,18 +334,89 @@ function M.find(names, opts)
   return matches
 end
 
+--- Split a Windows path into a prefix and a body, such that the body can be processed like a POSIX
+--- path. Path must use forward slashes (/) as the path separator and must not have any extraneous
+--- slashes.
+---
+--- Does not check if the path is a valid Windows path. Invalid paths will give invalid results.
+---
+--- Examples:
+--- - `//./C:/foo/bar` -> `//./C:`, `/foo/bar`
+--- - `//?/UNC/server/share/foo/bar` -> `//?/UNC/server/share`, `/foo/bar`
+--- - `C:/foo/bar` -> `C:`, `/foo/bar`
+--- - `C:foo/bar` -> `C:`, `foo/bar`
+---
+--- @param path string Path to split.
+--- @return string, string : prefix, body
+local function split_windows_path(path)
+  local prefix = ''
+
+  local function process_unc_path()
+    local server_share = path:match('[^/]+/[^/]+/')
+
+    if server_share == nil then
+      error('Invalid Windows UNC path')
+    end
+
+    prefix = prefix .. server_share --[[ @as string ]]
+    path = path:sub(#server_share + 1)
+  end
+
+  -- Device paths
+  if path:match('^//[?.]/') then
+    prefix = path:sub(1, 4)
+    path = path:sub(5)
+
+    local device = path:match('[^/]+/')
+
+    if device == nil then
+      error('Invalid Windows device path')
+    end
+
+    prefix = prefix .. device --[[ @as string ]]
+    path = path:sub(#device + 1)
+
+    if device == 'UNC/' then
+      process_unc_path()
+    end
+  -- UNC paths
+  elseif vim.startswith(path, '//') and not vim.startswith(path, '///') then
+    prefix = '//'
+    path = path:sub(3)
+    process_unc_path()
+  -- Drive paths
+  elseif path:match('^%w:') then
+    prefix = path:sub(1, 2)
+    path = path:sub(3)
+  end
+
+  -- If there is a slash at the end of the prefix, move it to the start of the body.
+  if prefix:sub(-1) == '/' then
+    prefix = prefix:sub(1, -2)
+    path = '/' .. path
+  end
+
+  return prefix, path
+end
+
 --- @class vim.fs.normalize.Opts
 --- @inlinedoc
 ---
 --- Expand environment variables.
 --- (default: `true`)
 --- @field expand_env boolean
 
---- Normalize a path to a standard format. A tilde (~) character at the
---- beginning of the path is expanded to the user's home directory and
---- environment variables are also expanded. `.` and `..` components are also
---- resolved.
+--- Normalize a path to a standard format. A tilde (~) character at the beginning of the path is
+--- expanded to the user's home directory and environment variables are also expanded. `.` and `..`
+--- components are also resolved, except when the path is relative and trying to resolve it would
+--- result in an absolute path. For example: the following will not get resolved:
+--- - `.` as the only part in a relative path (e.g., `.`, `././`).
+--- - Resolving `..` in a relative path leads outside the current directory (e.g., `../../foo`).
 ---
+--- NOTE: Attempting to resolve `..` past the root directory in POSIX or past the root of a drive in
```

Comment:
This doesn’t need to be an error. In the root directory, `..` and `.` are equivalent (ie they both point back to the root directory), so this should be valid.